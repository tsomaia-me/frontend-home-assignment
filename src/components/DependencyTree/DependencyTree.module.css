.root {
    margin: 0;
    padding: 0;
    list-style-type: none;
}

.root > li {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    cursor: default;
    padding: 8px 16px;
}

.root > li:hover {
    background-color: #e0e0e0;
}

.root > li > .left {
    display: flex;
    flex-direction: row;
    align-items: center;
}

.spinner {
    animation: spin 1s linear infinite;
}

@keyframes spin {
    to {
        transform: rotate(360deg);
    }
}
