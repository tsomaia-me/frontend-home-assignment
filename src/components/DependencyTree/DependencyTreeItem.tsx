import React from 'react'
import styles from './DependencyTree.module.css'

export interface DependencyTreeItemProps extends React.ComponentProps<'li'> {
  isExpanded?: boolean
  isLoading?: boolean
}

function DependencyTreeItem({ isExpanded, isLoading, children, ...restProps }: DependencyTreeItemProps) {
  return (
    <li {...restProps}>
      <div className={styles.left}>
        <span className="material-icons">{isExpanded ? 'expand_less' : 'expand_more'}</span>
        <span>{children}</span>
      </div>
      {isLoading && (
        <div className={styles.spinner}>
          <span className="material-icons">sync</span>
        </div>
      )}
    </li>
  )
}

export default DependencyTreeItem
