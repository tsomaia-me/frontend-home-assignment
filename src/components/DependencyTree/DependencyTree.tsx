import React, { useCallback, useEffect, useState } from 'react'
import { observer } from 'mobx-react'

import styles from './DependencyTree.module.css'
import { usePackageStore } from '../../hooks/usePackageStore'
import DependencyTreeItem from './DependencyTreeItem'

export interface DependencyTreeProps {
  dependencies: Array<string>
  depth?: number
  isAllExpanded?: boolean
}

export interface DependencyItemProps {
  packageName: string
  depth: number
  isAllExpanded: boolean
}

function DependencyTree({ dependencies, depth = 0, isAllExpanded = false }: DependencyTreeProps) {
  return (
    <ul className={styles.root}>
      {dependencies.map(packageName => (
        <DependencyItem
          key={packageName}
          packageName={packageName}
          depth={depth}
          isAllExpanded={isAllExpanded}
        />
      ))}
    </ul>
  )
}

const DependencyItem = observer(({ packageName, isAllExpanded, depth }: DependencyItemProps) => {
  const packageStore = usePackageStore()
  const [isExpanded, setIsExpanded] = useState(false)
  const toggleExpansion = useCallback(() => setIsExpanded(value => !value), [])
  const packageDetails = packageStore.packages[packageName]

  useEffect(() => {
    if (isExpanded) {
      packageStore.loadPackageDetails(packageName)
    }
  }, [packageName, isExpanded, packageStore])

  useEffect(() => {
    if (isAllExpanded) {
      setIsExpanded(true)
    }
  }, [isAllExpanded])

  return (
    <>
      <DependencyTreeItem
        isExpanded={isExpanded}
        isLoading={isExpanded && !packageDetails}
        style={{ paddingLeft: 8 + (24 * depth) }}
        children={packageName}
        onClick={toggleExpansion}
      />

      {isExpanded && packageDetails && (
        <DependencyTree
          dependencies={Object.keys(packageDetails.package.dependencies ?? {})}
          depth={depth + 1}
          isAllExpanded={isAllExpanded}
        />
      )}
    </>
  )
})

export default DependencyTree
