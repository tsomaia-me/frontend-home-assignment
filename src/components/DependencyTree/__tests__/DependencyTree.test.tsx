import React from 'react'
import renderer from 'react-test-renderer'
import DependencyTree from '../DependencyTree'

test('renders DependencyTree correctly', () => {
  const json = renderer.create(<DependencyTree dependencies={[]}/>).toJSON()
  expect(json).toMatchSnapshot()
})
