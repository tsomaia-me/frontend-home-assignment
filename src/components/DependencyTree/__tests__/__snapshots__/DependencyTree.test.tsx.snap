// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DependencyTree correctly 1`] = `
<ul
  className="root"
/>
`;
